pkg/skaffold/build/build.go

```diff
@@ -30,14 +30,20 @@ type Artifact struct {
 	Tag       string
 }
 
+type Result struct {
+	Target *latest.Artifact
+	Result *Artifact
+	Error  error
+}
+
 // Builder is an interface to the Build API of Skaffold.
 // It must build and make the resulting image accesible to the cluster.
 // This could include pushing to a authorized repository or loading the nodes with the image.
 // If artifacts is supplied, the builder should only rebuild those artifacts.
 type Builder interface {
 	Labels() map[string]string
 
-	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Result, error)
```

Comment:
The `Artifact` is used in Build output Template flag, `--output`.  By default it returns "{{.}}" which is the Artifact Object.  There should be a corresponding change here https://github.com/GoogleContainerTools/skaffold/blob/master/cmd/skaffold/app/cmd/build.go#L93 

However, before we do that, i left comment on the design doc to use `go channels` which can mitigate the issue where build  events get handled as they complete.

Reply:
re: using channels for the build results:

I'm thinking that this might not be a visible enough issue to warrant blocking this PR on. I agree it would be the "right" way to handle sending results back to the runner, but I think it's an enhancement that could be a follow up PR, as most users will probably not notice the delay in events/state changes (even cloud code users). WDYT?

Reply:
yes i agree. This is a good change and irrelevant to how we decide to send events.

Reply:
We decided to go with channels. 
Resolving this.